cmd/prysmctl/validator/withdraw.go

```diff
@@ -0,0 +1,123 @@
+package validator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/logrusorgru/aurora"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/client/beacon"
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	log "github.com/sirupsen/logrus"
+	"github.com/urfave/cli/v2"
+	"go.opencensus.io/trace"
+)
+
+func setWithdrawalAddresses(c *cli.Context) error {
+	ctx, span := trace.StartSpan(c.Context, "withdrawal.setWithdrawalAddress")
+	defer span.End()
+	au := aurora.NewAurora(true)
+	beaconNodeHost := c.String(BeaconHostFlag.Name)
+	if !c.IsSet(PathFlag.Name) {
+		return fmt.Errorf("no --%s flag value was provided", PathFlag.Name)
+	}
+	foundFilePaths, err := findWithdrawalFiles(c.String(PathFlag.Name))
+	if err != nil {
+		return errors.Wrap(err, "failed to find withdrawal files")
+	}
+	setWithdrawalAddressJsons := make([]*apimiddleware.SignedBLSToExecutionChangeJson, 0)
+	for _, foundFilePath := range foundFilePaths {
+		b, err := os.ReadFile(filepath.Clean(foundFilePath))
+		if err != nil {
+			return errors.Wrap(err, "failed to open file")
+		}
+		var to []*apimiddleware.SignedBLSToExecutionChangeJson
+		if err := json.Unmarshal(b, &to); err != nil {
+			log.Warnf("provided file: %s, is not a list of signed withdrawal messages", foundFilePath)
+			continue
+		}
+		setWithdrawalAddressJsons = append(setWithdrawalAddressJsons, to...)
+	}
+	if len(setWithdrawalAddressJsons) == 0 {
+		return errors.New("the list of signed requests is empty")
+	}
+	for _, request := range setWithdrawalAddressJsons {
+		fmt.Println("SETTING VALIDATOR INDEX " + au.Red(request.Message.ValidatorIndex).String() + " TO WITHDRAWAL ADDRESS " + au.Red(request.Message.ToExecutionAddress).String())
+	}
+	return callWithdrawalEndpoints(ctx, beaconNodeHost, setWithdrawalAddressJsons)
+}
+
+func callWithdrawalEndpoints(ctx context.Context, host string, request []*apimiddleware.SignedBLSToExecutionChangeJson) error {
+	client, err := beacon.NewClient(host)
+	if err != nil {
+		return err
+	}
+	if err := client.SubmitChangeBLStoExecution(ctx, request); err != nil {
+		return err
+	}
+	log.Infof("Successfully published messages to update %d withdrawal addresses.", len(request))
+	return checkIfWithdrawsAreInPool(ctx, client, request)
+}
+
+func checkIfWithdrawsAreInPool(ctx context.Context, client *beacon.Client, request []*apimiddleware.SignedBLSToExecutionChangeJson) error {
+	log.Info("verifying requested withdrawal messages known to node...")
+	poolResponse, err := client.GetBLStoExecutionChanges(ctx)
+	if err != nil {
+		return err
+	}
+	missingInPool := false
+	for _, w := range request {
```

Comment:
This search O(N^2)... I'm not sure how important this is but this may be a large list of checks. I have no good intuition whether we want to change this or not.

Reply:
In addition to being n^2, I think this is just a confusing way to write this loop - why is this even using binary search? An easy and clear implementation would be:
- build a map of the all the address+index pairs you're searching for
- loop once over the pool, check the map for each given pair, and delete any match from the map
- any pair left in the map at the end is not in the pool

Reply:
the sad part is that we already have a map in the beacon with the potentially large slice, if we eventually add an endpoint to check presence of a particular address, this could be done easily

Reply:
I'll refactor based on kasey's comments for the current implementation anyways + add something for checking the length of the pool anyways.